pkg/controller/container/cluster.go

```diff
@@ -0,0 +1,530 @@
+// and Start it when the Manager is Started.
+func (c *GKEClusterController) SetupWithManager(mgr ctrl.Manager) error {
+	r := resource.NewManagedReconciler(mgr,
+		resource.ManagedKind(v1beta1.GKEClusterGroupVersionKind),
+		resource.WithExternalConnecter(&clusterConnector{kube: mgr.GetClient(), newServiceFn: container.NewService}))
+
+	name := strings.ToLower(fmt.Sprintf("%s.%s", v1beta1.GKEClusterKindAPIVersion, v1beta1.Group))
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1beta1.GKECluster{}).
+		Complete(r)
+}
+
+type clusterConnector struct {
+	kube         client.Client
+	newServiceFn func(ctx context.Context, opts ...option.ClientOption) (*container.Service, error)
+}
+
+func (c *clusterConnector) Connect(ctx context.Context, mg resource.Managed) (resource.ExternalClient, error) {
+	i, ok := mg.(*v1beta1.GKECluster)
+	if !ok {
+		return nil, errors.New(errNotCluster)
+	}
+
+	p := &gcpv1alpha3.Provider{}
+	if err := c.kube.Get(ctx, meta.NamespacedNameOf(i.Spec.ProviderReference), p); err != nil {
+		return nil, errors.Wrap(err, errGetProvider)
+	}
+
+	s := &corev1.Secret{}
+	n := types.NamespacedName{Namespace: p.Spec.Secret.Namespace, Name: p.Spec.Secret.Name}
+	if err := c.kube.Get(ctx, n, s); err != nil {
+		return nil, errors.Wrap(err, errGetProviderSecret)
+	}
+
+	client, err := c.newServiceFn(ctx,
+		option.WithCredentialsJSON(s.Data[p.Spec.Secret.Key]),
+		option.WithScopes(container.CloudPlatformScope))
+	return &clusterExternal{cluster: client, projectID: p.Spec.ProjectID, kube: c.kube}, errors.Wrap(err, errNewClient)
+}
+
+type clusterExternal struct {
+	kube      client.Client
+	cluster   *container.Service
+	projectID string
+}
+
+func (e *clusterExternal) Observe(ctx context.Context, mg resource.Managed) (resource.ExternalObservation, error) { // nolint:gocyclo
+	cr, ok := mg.(*v1beta1.GKECluster)
+	if !ok {
+		return resource.ExternalObservation{}, errors.New(errNotCluster)
+	}
+
+	existing, err := e.cluster.Projects.Locations.Clusters.Get(gke.GetFullyQualifiedName(e.projectID, cr.Spec.ForProvider)).Context(ctx).Do()
+	if err != nil {
+		return resource.ExternalObservation{}, errors.Wrap(resource.Ignore(gcp.IsErrorNotFound, err), errGetCluster)
+	}
+
+	cr.Status.AtProvider = gke.GenerateObservation(*existing)
+	currentSpec := cr.Spec.ForProvider.DeepCopy()
+	gke.LateInitializeSpec(&cr.Spec.ForProvider, *existing)
+	if !cmp.Equal(currentSpec, &cr.Spec.ForProvider) {
+		if err := e.kube.Update(ctx, cr); err != nil {
+			return resource.ExternalObservation{}, errors.Wrap(err, errManagedUpdateFailed)
+		}
+	}
+
+	switch cr.Status.AtProvider.Status {
+	case v1beta1.ClusterStateRunning:
+		cr.Status.SetConditions(v1alpha1.Available())
+		resource.SetBindable(cr)
+	case v1beta1.ClusterStateProvisioning:
+		cr.Status.SetConditions(v1alpha1.Creating())
+	case v1beta1.ClusterStateUnspecified, v1beta1.ClusterStateDegraded, v1beta1.ClusterStateError, v1beta1.ClusterStateReconciling:
```

Comment:
Is this cluster really unavailable in all of these states? What does reconciling mean?

Reply:
Reconciling is the status that is returned when some sort of update is being performed.